From 9ac77b41d3ce82a8f5d2fed1f8bdae5db5134f72 Mon Sep 17 00:00:00 2001
From: Dan Kephart <dan.kephart@lairdtech.com>
Date: Thu, 5 Feb 2015 14:16:39 -0500
Subject: [PATCH] laird: ath6kl: change GFP_KERNEL to GFP_ATOMIC in nlatheros
 event multicast

Use of GFP_KERNEL in an interrupt context in ath6kl_drv_event_multicast()
and ath6kl_wmi_event_multicast() could result in a kernel crash.  This is
due to GFP_KERNEL allocations being are allowed to sleep until a page is
available.

Bug: 6950
---
 drivers/net/wireless/ath/ath6kl/wmi.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/net/wireless/ath/ath6kl/wmi.c b/drivers/net/wireless/ath/ath6kl/wmi.c
index 5d26d1aed22f..b0e25589c304 100644
--- a/drivers/net/wireless/ath/ath6kl/wmi.c
+++ b/drivers/net/wireless/ath/ath6kl/wmi.c
@@ -4559,7 +4559,7 @@ static void ath6kl_wmi_event_multicast(enum wmi_cmd_id cmd_id, u8 *datap, int le
     struct sk_buff *msg;
 	void *hdr;
 
-    msg = genlmsg_new(NLMSG_DEFAULT_SIZE, GFP_KERNEL);
+    msg = genlmsg_new(NLMSG_DEFAULT_SIZE, GFP_ATOMIC);
 	if (!msg)
 		return;
 
@@ -4575,7 +4575,7 @@ static void ath6kl_wmi_event_multicast(enum wmi_cmd_id cmd_id, u8 *datap, int le
 
 	genlmsg_end(msg, hdr);
 
-	genlmsg_multicast(&atheros_fam, msg, 0, ATH_MCGRP_EVENTS, GFP_KERNEL);
+	genlmsg_multicast(&atheros_fam, msg, 0, ATH_MCGRP_EVENTS, GFP_ATOMIC);
 
 	return;
 
@@ -4589,7 +4589,7 @@ void ath6kl_drv_event_multicast(enum atheros_cmd_id cmd_id, unsigned int reason)
 	struct sk_buff *msg = NULL;
 	void *hdr = NULL;
 
-	msg = genlmsg_new(NLMSG_DEFAULT_SIZE, GFP_KERNEL);
+	msg = genlmsg_new(NLMSG_DEFAULT_SIZE, GFP_ATOMIC);
 	if (msg)
 		hdr = genlmsg_put(msg, 0, 0, &atheros_fam, 0, cmd_id);
 	if (!hdr)
@@ -4598,7 +4598,7 @@ void ath6kl_drv_event_multicast(enum atheros_cmd_id cmd_id, unsigned int reason)
 		if(nla_put_u32(msg, ATHEROS_ATTR_MSG, reason))
 			goto nla_put_failure;
 		genlmsg_end(msg, hdr);
-		genlmsg_multicast(&atheros_fam, msg, 0, ATH_MCGRP_EVENTS, GFP_KERNEL);
+		genlmsg_multicast(&atheros_fam, msg, 0, ATH_MCGRP_EVENTS, GFP_ATOMIC);
 	}
 
 	return;
-- 
2.11.0

